### PR TITLE
[EDMTP-434] Create a custom injector for Child Resource(s)

### PR DESCRIPTION
**Pull request by rskaldin**

### CODE REVIEW

**Comment by rskaldin** [comment]
I have read the CLA Document and I hereby sign the CLA

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java [hunk 1]
```diff
@@ -140,4 +148,99 @@ private static boolean isValidObjectType(Class<?> value, Class<?>... allowedType
...
+    }
+
+    /**
+     * Retrieves the last name from a given relative path
+     * @param relativePath {@code String} relative path
+     * @return {@code String} that representing the last name
+     */
+    public static String getLastNodeName(String relativePath) {
+        String lastNodeName = relativePath;
+        if (relativePath.endsWith("/")) {
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java:241]: `CoreConstants.SEPARATOR_SLASH`

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/filters/NonNullPredicate.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Objects;
+import java.util.function.Predicate;
+
+import org.apache.sling.api.resource.Resource;
+
+/**
+ * The predicate that checks if a resource is non-null
+ */
+public class NonNullPredicate implements Predicate<Resource> {
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/aem/toolkit/core/injectors/filters/NonNullPredicate.java:24]: @all I still believe we must unite these predicates. or create a third one. How much is the probability that a user needs nulls but does not need ghosts?) we are fighting with boilerplate via all of this code,  but we bring more...

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/filters/NoGhostPredicate.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Evaluates this predicate on the given argument.
+     *
+     * @param resource the input argument
+     * @return {@code true} if the input argument matches the predicate,
+     * otherwise {@code false}
+     */
+    @Override
+    public boolean test(Resource resource) {
+        return !resource.isResourceType("wcm/msm/components/ghost");
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/aem/toolkit/core/injectors/filters/NoGhostPredicate.java:34]: NPE-prone

File: core/src/test/java/com/exadel/aem/toolkit/core/injectors/ChildInjectorTest.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        context.addModelsForClasses(TestModelChild.class);
+        context.registerInjectActivateService(new ChildInjector());
+        context.load().json("/com/exadel/aem/toolkit/core/injectors/childInjector.json", "/content");
+        ResourceResolver resolver = context.resourceResolver();
+        context.request().setResource(resolver.getResource("/content/jcr:content"));
+        testModel = context.request().adaptTo(TestModelChild.class);
+    }
+
+    @Test
+    public void testInjectChild() {
```
Comment (smiakchilo) [inline core/src/test/java/com/exadel/aem/toolkit/core/injectors/ChildInjectorTest.java:54]: (naming) Please note that all other test cases in `core` are named like `shouldDoSomething`

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/ChildrenInjector.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+     * @param type            The {@code Type} to adapt to
+     * @param predicates      {@code List<Predicate<Resource>>} of predicates
+     * @return {@code List<Object>} of filtered and adapted objects. Otherwise, empty list is returned
+     */
+    private List<Object> getFilteredList(Resource currentResource, Type type, List<Predicate<Resource>> predicates) {
+        if (currentResource == null) {
+            return Collections.emptyList();
+        }
+
+        final Class<?> actualType = InjectorUtils.getActualType((ParameterizedType) type);
```
[the remark this answers is not in the document]
Comment (rskaldin) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/ChildrenInjector.java:168]: Yes, I suppose. A variable used in lambda expression should be a final or effectively final.

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java [hunk 6]
```diff
@@ -140,4 +148,99 @@ private static boolean isValidObjectType(Class<?> value, Class<?>... allowedType
         }
         return Arrays.asList(allowedTypes).contains(value) || value.equals(Object.class);
     }
+
+    /**
+     * Retrieves whether the provided {@code Type} of a Java class member is a parametrized collection
+     * @param type {@code Type}
+     * @return True of false
+     */
+    public static boolean isTypeCollection(Type type) {
```
[the remark this answers is not in the document]
Comment (rskaldin) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java:162]: Thank you, fixed!

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java [hunk 7]
```diff
@@ -140,4 +148,99 @@ private static boolean isValidObjectType(Class<?> value, Class<?>... allowedType
...
+        return type instanceof ParameterizedType
+            && ClassUtils.isAssignable((Class<?>) ((ParameterizedType) type).getRawType(), Collection.class);
+    }
+
+    /**
+     * Retrieves actual type parameter from parameterized type
+     * @param parameterizedType {@code Class} that represents a parameterized type
+     * @return {@code Class} representing the actual type arguments to this type
+     */
+    public static Class<?> getActualType(ParameterizedType parameterizedType) {
```
[the remark this answers is not in the document]
Comment (rskaldin) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java:172]: Thank you, renamed!

File: core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java [hunk 8]
```diff
@@ -140,4 +148,99 @@ private static boolean isValidObjectType(Class<?> value, Class<?>... allowedType
...
+     * @return <T> initialized object instance
+     */
+    public static <T> T getObjectInstance(Class<? extends T> instanceClass) {
+        try {
+            return instanceClass.getConstructor().newInstance();
+        } catch (InstantiationException
+            | IllegalAccessException
+            | InvocationTargetException
+            | NoSuchMethodException ex) {
+            LOG.error("Could not initialize object " + instanceClass.getName(), ex);
```
[the remark this answers is not in the document]
Comment (rskaldin) [reply core/src/main/java/com/exadel/aem/toolkit/core/injectors/InjectorUtils.java:189]: Fixed, thanks!

**Pull request merged**